streaming/base/format/base/reader.py

```diff
@@ -63,6 +64,145 @@ def __len__(self) -> int:
+        # the shard raw as not present, in order to achieve consistency.
+        if not raw_files_present:
+            is_raw_present = False
+        elif raw_files_present < len(self.file_pairs):
+            is_raw_present = False
+            for raw_info, _ in self.file_pairs:
+                if raw_info:
+                    filename = os.path.join(self.dirname, self.split, raw_info.basename)
+                    if filename in ls:
+                        os.remove(filename)
+        else:
+            is_raw_present = True
+
+        # Same as the above, but for shard zip files.
+        if not zip_files_present:
+            is_zip_present = False
+        elif zip_files_present < len(self.file_pairs):
+            is_zip_present = False
+            for _, zip_info in self.file_pairs:
+                if zip_info:
+                    filename = os.path.join(self.dirname, self.split, zip_info.basename)
+                    if filename in ls:
+                        os.remove(filename)
+        else:
+            is_zip_present = True
+
+        # Do we keep_zip?
+        if keep_zip:
+            # If we can keep_zip, and we do, and have either raw or zip, we must have the other one
+            # too.
+            if self.compression and (is_zip_present ^ is_raw_present):
+                if is_raw_present:
+                    is_raw_present = False
+                    for raw_info, _ in self.file_pairs:
+                        filename = os.path.join(self.dirname, self.split, raw_info.basename)
+                        os.remove(filename)
+                elif is_zip_present:
+                    is_zip_present = False
+                    for _, zip_info in self.file_pairs:
+                        filename = os.path.join(self.dirname, self.split, zip_info.basename)
+                        os.remove(filename)
+        else:
+            # If we don't keep_zip, drop any zip files.
+            if is_zip_present:
+                is_zip_present = False
+                for _, zip_info in self.file_pairs:
+                    filename = os.path.join(self.dirname, self.split, zip_info.basename)
+                    os.remove(filename)
+
+        # Now, the shard is either entirely or not at all present given keep_zip.
+        return is_raw_present
+
+    def get_raw_size(self) -> int:
+        """Get the raw (uncompressed) size of this shard.
+
+        Returns:
+            int: Size in bytes.
+        """
+        size = 0
+        for info, _ in self.file_pairs:
+            size += info.bytes
+        return size
+
+    def get_zip_size(self) -> Optional[int]:
+        """Get the zip (compressed) size of this shard, if compression was used.
+
+        Returns:
+            Optional[int]: Size in bytes, or ``None`` if does not exist.
+        """
+        size = 0
+        for _, info in self.file_pairs:
+            if info is None:
```

Comment:
i guess it is guaranteed that if one `info` is `None`, all `info` are `None`?

Reply:
yep, although not sure where or how to note this to the reader